drgn_tools/meminfo.py

```diff
@@ -0,0 +1,755 @@
+    global _MAX_NUM_NODES, _MAX_NUM_ZONES
+    global _PAGE_SIZE, _PAGE_SHIFT, _PMD_SHIFT, _HPAGE_SHIFT
+    global _HPAGE_PMD_SHIFT, _HPAGE_PMD_ORDER, _HPAGE_PMD_NR
+    global _VMALLOC_END, _VMALLOC_START, _VMALLOC_SPACE
+
+    uts = get_uts(prog)
+    # Set UEK version
+    if "4.14.35" in uts["release"]:
+        _UEK_VERSION = 5
+    elif "5.4.17" in uts["release"]:
+        _UEK_VERSION = 6
+    else:
+        _UEK_VERSION = 7
+
+    # Determine page size
+    _PAGE_SIZE = prog.constant("PAGE_SIZE").value_()
+    _PAGE_SHIFT = prog.constant("PAGE_SHIFT").value_()
+
+    # Set arch
+    if "aarch64" in uts["machine"]:
+        _IS_AARCH_64 = True
+
+        # include/linux/numa.h
+        _NODES_SHIFT = 2
+
+        # arch/arm64/include/asm/pgtable-hwdef.h
+        n = 2
+        _PMD_SHIFT = (_PAGE_SHIFT - 3) * (4 - (n)) + 3
+        n = 1
+        PGDIR_SHIFT = (_PAGE_SHIFT - 3) * (4 - (n)) + 3
+
+        # include/asm-generic/pgtable-nopud.h
+        P4D_SHIFT = PGDIR_SHIFT
+        PUD_SHIFT = P4D_SHIFT
+        PUD_SIZE = 1 << PUD_SHIFT
+
+        # arch/arm64/include/asm/pgtable.h
+        _HPAGE_SHIFT = _PMD_SHIFT
+
+        SZ_64K = 0x00010000
+        SZ_128M = 0x08000000
+        VA_BITS = 48
+        VA_START = 0xFFFFFFFFFFFFFFFF - (1 << VA_BITS) + 1
+        PAGE_OFFSET = 0xFFFFFFFFFFFFFFFF - (1 << (VA_BITS - 1)) + 1
+        STRUCT_PAGE_MAX_SHIFT = 6
+        KASAN_SHADOW_SIZE = 0
+        MODULES_VADDR = VA_START + KASAN_SHADOW_SIZE
+        MODULES_VSIZE = SZ_128M
+        MODULES_END = MODULES_VADDR + MODULES_VSIZE
+        VMEMMAP_SIZE = 1 << (VA_BITS - _PAGE_SHIFT - 1 + STRUCT_PAGE_MAX_SHIFT)
+        _VMALLOC_START = MODULES_END
+        _VMALLOC_END = PAGE_OFFSET - PUD_SIZE - VMEMMAP_SIZE - SZ_64K
+
+    else:
+        _IS_X86_64 = True
+
+        # include/linux/numa.h
+        _NODES_SHIFT = 10
+
+        # arch/x86/include/asm/pgtable_64_types.h
+        _PMD_SHIFT = 21
+
+        VMALLOC_SIZE_TB = 32
+        _VMALLOC_START = prog["vmalloc_base"].value_()
+        _VMALLOC_END = _VMALLOC_START + (VMALLOC_SIZE_TB << 40) - 1
+
+        # arch/x86/include/asm/page_types.h
+        _HPAGE_SHIFT = _PMD_SHIFT
+
+    _MAX_NUM_NODES = 1 << _NODES_SHIFT
+    _MAX_NUM_ZONES = prog["__MAX_NR_ZONES"].value_()
+
+    _HPAGE_PMD_SHIFT = _PMD_SHIFT
+    _HPAGE_PMD_ORDER = _HPAGE_PMD_SHIFT - _PAGE_SHIFT
+    _HPAGE_PMD_NR = 1 << _HPAGE_PMD_ORDER
+
+
+def get_active_numa_nodes(prog: Program) -> List[Object]:
+    """
+    Get a list of all active NUMA nodes.
+
+    :returns: a list of ``struct pglist_data *`` objects.
+    """
+    try:
+        active_numa_nodes = prog.cache["active_numa_nodes"]
+    except KeyError:
+        active_numa_nodes = []
+        node_data = prog["node_data"]
+        for i in range(_MAX_NUM_NODES):
+            if node_data[i].value_() != 0x0:
+                active_numa_nodes.append(node_data[i])
+        prog.cache["active_numa_nodes"] = active_numa_nodes
+    return active_numa_nodes
```

Comment:
Please store everything for this module within a cache dictionary named `drgn_tools.meminfo`. EG:
```suggestion
    try:
        cache = prog.cache.setdefault("drgn_tools.meminfo", {})
        active_numa_nodes = cache["active_numa_nodes"]
    except KeyError:
        active_numa_nodes = []
        node_data = prog["node_data"]
        for i in range(_MAX_NUM_NODES):
            if node_data[i].value_() != 0x0:
                active_numa_nodes.append(node_data[i])
        cache["active_numa_nodes"] = active_numa_nodes
    return active_numa_nodes
```

Reply:
Sounds good. Updated.